Ensure Correct COT Selection in QB When Sharing Formatters

Areyes42 opened the pull request:
Fixes #5503

<!--
> [!WARNING]  
> This PR affects database migrations. See [migration testing instructions](https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration).
-->

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add automated tests
- [x] Add relevant issue to release milestone
- [x] Add relevant documentation (Tester - Dev)

### Testing instructions

1. Create two CollectionObjectTypes with the same Catalog Number Format Name which differs from that of the Collection's default Catalog Number format
2. Create a new Collection Object query, add Catalog Number to the query, and switch the type from `Any` to `Equal`
3. Try to select both COTs
- [ ] Verify that it selects the proper COT regardless of the Catalog Number format

sharadsw commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/Formatter.tsx:
```diff
@@ -140,9 +140,24 @@ function FormatSelect({
+  const compositeValue = React.useMemo(() => {
+    if (availableFormatters && availableFormatters.length > 0) {
+      if (selectedIndex !== null && availableFormatters[selectedIndex]) {
+        return `${availableFormatters[selectedIndex].name}|||${selectedIndex}`;
```
Since this string format is used in multiple places, can you make a function that accepts two params `name` and `index` and returns this generated string? It would help not only with reusability but also to document what the string represents

sharadsw commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/Formatter.tsx:
```diff
@@ -173,13 +188,28 @@ function FormatSelect({
-            onValueChange={handleChange}
+            value={compositeValue}
+            onValueChange={(value) => {
+              const parts = value.split('|||');
```
In a similar vein, can you make `|||` a constant and reuse that here and in the function you create? It would help document the purpose of the string and keep any future changes more isolated.

sharadsw reviewed (changes requested):
I like your approach. Functionally your code looks good, just pointing out some minor code smells.

Areyes42 commented:
> I like your approach. Functionally your code looks good, just pointing out some minor code smells.

Good point, I've added some helpers to separate the logic and hopefully make it a little more intuitive. Let me know if you think this looks better!

sharadsw commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/Formatter.tsx:
```diff
@@ -173,13 +203,26 @@ function FormatSelect({
+            value={compositeValue}
+            onValueChange={(value) => {
+              const parsed = parseFormatterValue(value);
+              if (parsed) {
+                setSelectedIndex(parsed.index);
+                if (handleChange) {
+                  handleChange(parsed.name);
+                }
+              } else {
+                setSelectedIndex(null);
+                if (handleChange) {
+                  handleChange(value);
+                }
+              }
```
(optional) some shorthands if you want to make it cleaner
```suggestion
              const index = parsed?.index ?? null;
              const name = parsed?.name ?? value;
              setSelectedIndex(index);
              handleChange?.(name);
```

sharadsw reviewed (approved):
Looks good, nice work!

lexiclevenger reviewed (approved):
- [x] Verify that it selects the proper COT regardless of the Catalog Number format

---

Looks good!


https://github.com/user-attachments/assets/ed8a1aae-88b1-4029-a8b4-6d12a72e2531

pashiav reviewed (approved):
- [X] Verify that it selects the proper COT regardless of the Catalog Number format
---
Works as expected!

melton-jason reviewed (commented):
Nice work on this! I like the implementation; pretty clever idea! 🚀 

I will mention that while the CollectionObjectType the user selects while the page is loaded will always be respected, it will always revert to the first CollectionObjectType in the list with the same format when re-rendered!

https://github.com/user-attachments/assets/3825793e-ca5c-45b4-89b6-675841d2215a

This is because only the format name (nothing of the CollectionObjectType) is saved to the `SpQueryField` in the database; when the page is rendered initially, it isn't formatted with the `formatterSeparator` and `index` and is just the name of the formatter itself. 
So Specify has no clue which CollectionObjectType it can be referring to. 

There's also a somewhat smaller Issue were changing the CollectionObjectType to another which shares the same format doesn't trigger a change on the field (because nothing has actually changed on the SpQueryField!) 

https://github.com/user-attachments/assets/9a72cc65-2823-4dcf-9dbe-7f65ec0f1a94

---

We're relatively having a lot of Issues regarding this feature (no thanks to my initial implementation in the first place!), and I think that's usually an indicator to take a step back and re-evaluate the core problem and potential solutions. 

@specify/ux-testing, @specify/dev-testing 

Instead of having a separate option for each CollectionObjectType, could we instead have an option for each Format and have the title of the option be a combination of all the COT names with that format? 

It could look something like this: 

<img width="1439" alt="Screenshot 2025-02-26 at 9 30 38 PM" src="https://github.com/user-attachments/assets/f954c726-def0-409a-aa62-a3e4aa437dbe" />


Or we can go even more descriptive to try and be as clear as possible: the user _isn't_ selecting/filtering on COT, they're only applying a different formatter for the field: 

<img width="1439" alt="Screenshot 2025-02-26 at 9 54 29 PM" src="https://github.com/user-attachments/assets/0d438d7a-b7e3-4fb9-91e6-ba6c1d3f0d3e" />

Alternatively, we could move away from the `SpQueryField -> FormatName` approach entirely and come up with something different which would allow the user to select an individual COT name.

emenslin commented:
I agree with Jasons comment and I think that grouping them together might make the most sense since it does not necessarily matter which specific one you use. I also think since we don't really have the gear other places aside from when you have something mapped as 'formatter' or 'aggregator' that adding 'Format As: ...' in this instance makes sense but we can go with something else if people have other opinions @specify/ux-testing

Areyes42 commented:
@melton-jason @emenslin 

Great suggestions! I've updated the logic to group COTs with identical formatters together. This should mitigate issues related to re-rendering and state changes in the form, since it no longer compares against the same formatter. Also, I've added the "Format As:" identifier to improve clarity, but we can adjust it if it seems too confusing.

emenslin reviewed (approved):
- [x] Verify that it selects the proper COT regardless of the Catalog Number format

Looks good! I ran into an issue where the query results were incorrect (I set the filter to equal and instead it showed all results), I couldn't recreate it but I would love if we could get one more testing review to verify that it is not an issue

pashiav reviewed (approved):
- [x] Verify that it selects the proper COT regardless of the Catalog Number format=
--- 

Looks good! 

@emenslin I didn't run into what you mentioned, it's working on my end!

melton-jason reviewed (commented):
Nice work!
Coming along nicely 👌
